### ISSUE TITLE
React Native support

### ISSUE THREAD

**Issue by Andreyco** [paypal/glamorous#7]
Did you considered supporting React Native?

If so, do you need any help?

**Comment by kentcdodds** [comment]
Happy to consider supporting React Native. I've never done anything with RN, so I wont be working on it personally, but happy to accept a PR for it :)

**Comment by Andreyco** [comment]
Thanks for quick response! 

I have feeling "anonymous" styled components are win-win regardless platform - naming is hard. Will try to work on this asap.

**Comment by ChristopherBiscardi** [comment]
I wonder if there is anything we should be doing in glamor to support this.

**Comment by Andreyco** [comment]
I don't think glamor has to do a single think. React Native support requires it's own components - `View`, `Text` instead of `Div` or `Span` for example.

Also, it would be great to limit properties to those known by React Native

**Comment by browniefed** [comment]
React Native uses `style` rather than `className`, and additionally `transform` is not a simple string it's an array of objects.
A lot of the things the glamor does don't necessarily apply to React Native or implementing it may be difficult. For example supporting media queries would require you to import `Dimensions` from react-native rather than `window.matchMedia`.
So it's not as simple as swapping over to use use View/Text instead of Div/Span

**Comment by ssomnoremac** [comment]
@browniefed , I agree it would almost be best to implement a separate simple implantation for native [like this ](https://gist.github.com/ssomnoremac/21ea4032c4bf5044e7fc34fe7f44a341) and work to align the implementation details rather than build on the same foundation -- although I really haven't looked at what's under the hood here. The RN implementation is really dead simple.

**Comment by kentcdodds** [comment]
Here's something relevant from @threepointone: https://gist.github.com/threepointone/ec7febc17f02fd91dcbabc087519088e
I think implementing glamor-react-native would be the first step for this.

**Comment by ssomnoremac** [comment]
To brownfields point, is the idea here to do the reverse of [react-native-web](https://github.com/necolas/react-native-web)? In other words, css/DOM -> RNstyles/RNcomponents? I think glamor for react-native is a much bigger project. Can we just have an implementation for native that doesn't use any css, just regular plain react-native styles on plain react-native components?

**Comment by kentcdodds** [comment]
> I think glamor for react-native is a much bigger project.

I'm not sure I follow, based on @threepointone's gist I linked above, it should be pretty straightforward... 🤔

**Comment by ssomnoremac** [comment]
I'm not really understanding the gist. Stylesheet is a helper in RN to allow you to construct className-like styles
```
let sheet = Stylesheet.create({
    red: {
        color: 'red'
    },
    blue: {
        color: 'blue'
    }
})

<Text styles={sheet.red}>Hello world!</Text>

```

You don't really need to use Stylesheet though. Simply using plain objects for styles is fine. 
```
let sheet = {
    red: {
        color: 'red'
...
```

so I don't understand why we would need to construct a Stylesheet for each component.

**Comment by threepointone** [comment]
Prevents the need to 'name' the styles, but still be performant (plain inline styles aren't, iiuc)

**Comment by browniefed** [comment]
The reason you want to use `StyleSheet` is that stylesheet will store the style in it's cache. That way every re-render you're not constantly recycling objects.

If you look at a `Stylesheet.create` return, the `styles.red` will be a number referencing the cache.
Now the part I haven't looked deeply into is the caching on the native side and when exactly and what is done to avoid calling FlattenStlyes and constantly passing objects over the bridge.

**Comment by browniefed** [comment]
If you look at create it's nothing special.
The main call is here `ReactNativePropRegistry.register(obj[key]);`

https://github.com/facebook/react-native/blob/master/Libraries/StyleSheet/StyleSheet.js#L201

**Comment by ssomnoremac** [comment]
I wasn't aware of this. Yeah, if caching can be accomplished without using Stylesheets that would be less hacky. I've also got a large RN app with a lot of inline styles 🙀

**Comment by statianzo** [comment]
Inspired by @ssomnoremac's suggestion, this is something I've been testing out:

https://gist.github.com/statianzo/c3b7c345abf82e833cb269dea7277709

It takes in the Div or Text component as instead of having a list predefined. A perk is if you wanted a limited amount of inheritance. Also, each styled component has its own Stylesheet.

**Comment by kentcdodds** [comment]
> It takes in the Div or Text component as instead of having a list predefined

Note that you can do this with glamorous as well:

```js
const MyDiv = glamorous('div')({ /* styles */ })
const MyThing = glamorous(AnyComponent)({ /* styles */ })
```

The predefined list is just for convenience:

```js
const MyDiv = glamorous.div({ /* styles */ })
```

**Comment by ssomnoremac** [comment]
I updated my [gist](https://gist.github.com/ssomnoremac/21ea4032c4bf5044e7fc34fe7f44a341) to use `ReactNativePropRegistry.register`. Correct me if I'm wrong, I don't need to memoize anything here as each `glamorous.View` (Styled.View in gist) usage will register it's own cached key. I'll try to actually test this soon.

**Comment by kentcdodds** [comment]
Cool, love to see it improving! Note that what you have there is still lacking quite a few features of `glamorous`:

1. [Pre-built `GlamorousComponent`s](https://github.com/paypal/glamorous#built-in-glamorouscomponents) (the jsxstyle inspired API)
2. [Dynamic styles](https://github.com/paypal/glamorous/blob/master/examples/README.md#dynamic--static-styles)
3. Pretty much everything that [glamor's API](https://github.com/threepointone/glamor) can help you with that you can't do with raw React Native stylesheets

But it's coming along! I think I'd really like to see it use something like `glamor-react-native` (yet to be created) under the hood.

**Comment by browniefed** [comment]
@ssomnoremac I would likely not use `ReactNativePropRegistry.register` directly. Probably not a great idea to depend on internals of RN yet. I'd say just use `Stylesheet.create` of a single style since it just calls `register` anyway.

I was just pointing to how things were being dealt with internally. People were concerned about calling `Stylesheet.create` multiple times. However with proper memoization and the caching that React Native already does we're probably fine.

I would say design an API that works then focus on making it fast and efficient. The performance of mobile apps is so heavily dependent on the phone and it's capabilities that premature optimization is probably not a great idea.

**Comment by ssomnoremac** [comment]
@browniefed, understood. To Kent's first point about this:

    <View backgroundColor="red" />

First, for a native `GamorousNativeComponent` we would need to create a `should-forward-native-props` function. Would it make sense to do the opposite, and only target possible styles to not-forward using a list like [this](https://github.com/vhpoet/react-native-styling-cheat-sheet). I'd imagine accounting for all props in a list for RN is close to impossible as the list would be very fluid. Second, it actually works out of the box as I've often [noticed](https://github.com/facebook/react-native/issues/12469) that style props will actually register as style properties in React Native. Although this is unintended, would this be such a bad thing to push to become as supported feature with the core team?

**Comment by browniefed** [comment]
As Brent had mentioned on that post we shouldn't depend on it. In all likely hood it will be removed in the future. RN is a fast moving target and if don't stick to the documented API we'll probably have to do a significant amount of refactoring in future.

Making this list isn't actually too bad. All properties are documented and typed, and are contained as prop types.

https://github.com/facebook/react-native/blob/909af08f24e3c65e46593a1f66d5484c34dda53c/Libraries/Components/View/ReactNativeStyleAttributes.js

**Comment by ssomnoremac** [comment]
I've got [something working](https://github.com/hyrwork/glamorous/blob/feature/react-native/src/native/index.js). I decided to align it to the current web glamorous `index.js` closure syntax. Styles are merged in a different way but as a stand-alone file I was able to drop it into my app and it worked. Example usage:

    const RedText = glamorous.Text({color:'red'})

    <RedText fontSize={24} >Hi there!</RedText>

Could use some help with the rest of the API and testing.

**Comment by kentcdodds** [comment]
Looks cool! I'm looking forward to seeing `glamorous-native` published soon <3

**Comment by Andreyco** [comment]
awesome. Will give it a shot and play a little with it tonight!

**Comment by Andreyco** [comment]
Looking at previous conversation, regarding caching... Having JSX style like API it would be pretty easy to cache styles on app level, instead of caching per `Stylesheet.create` calls.

**Comment by ssomnoremac** [comment]
@Andreyco , I just [changed that to ](https://github.com/hyrwork/glamorous/blob/feature/react-native/src/native/index.js)memoize it within the component itself. But on the app level would be better. Can you show how that would be done.

**Comment by Andreyco** [comment]
Nice! Sure, going to play & test it out

**Comment by ssomnoremac** [comment]
@Andreyco, just fixed a bug around `style` prop. I had 'styles`. Also removed `{props.children}` as not needed.

**Comment by ssomnoremac** [comment]
And wait, I don't think we need to memoize here, because we just have a class instance we can just lines 86-93 with just this
```
const cachedStylesNumber = this.cachedStylesNumber || StyleSheet.create({key: mergedStyles}).key
```

**Comment by browniefed** [comment]
Add a check if style is an array that's being passed in then spread it in. Last time I looked flatten styles won't do a deep flatten.

So if a user were to provide us an array of styles it wouldn't work.

However this might have changed

**Comment by ssomnoremac** [comment]
easy enough to do even if it has, thanks.

**Comment by ssomnoremac** [comment]
[updated](https://github.com/hyrwork/glamorous/blob/feature/react-native/src/native/index.js)

1. Simplified caching of static styles (ie. those not in `<View style={there}>`)
2. Checked for arrays in `{there}` and spread them if so
3. Added Text styles to the list of style names. 

Sorry @browniefed I couldn't find the exhaustive list from the link above. Perhaps when we get this properly incorporated we can fix that/ extract it to another file.

As for that, does the `/native/index.js` placement make sense to people?  If so, what's the next step. A proper PR?

**Comment by kentcdodds** [comment]
So, I'm thinking that this actually may be best as an entirely separate package in an entirely different repository. I don't think there would be very much code sharing (if any)... We could definitely link to one another's projects though. What do you all think?

**Comment by ssomnoremac** [comment]
Yeah, the problem is all of this stuff is all more simple in RN. We can leave it up to the community for a week to chime in. I'm open to starting and maintaining a new package if they agree.

**Comment by browniefed** [comment]
@ssomnoremac Also just realized. You should use `Animated`. Otherwise these components won't be able to be animated.
Swap `View` over to `Animated.View`. Anything that is not exported by `Animated` you can call `Animated.createAnimatedComponent()` on it.

**Comment by ssomnoremac** [comment]
@browniefed wouldn't having every component `Animated` add bloat? If so, would it make more sense to have a separate API for it like `glamAnimated.View({styles})` ?

**Comment by kentcdodds** [comment]
I would do: `glamorous.Animated.view` or something like that...

**Comment by browniefed** [comment]
Open to the API, the one that @kentcdodds just proposed would work.

But in reality no. Animated is just an HOC.
If it doesn't detect any Animated styles it doesn't do anything.

**Comment by kentcdodds** [comment]
Oh, also, note the special distinction in glamorous between capital and lower case properties. `glamorous.div` is a function that returns a component with styles. `glamorous.Div` is a component that accepts styles as props. Should probably be the same for glamorous-native to avoid confusing people...

**Comment by browniefed** [comment]
I would actually be hesitant to export anything name `Animated` from Glamorous.
This would then conflict with `import { Animated } from "react-native"`.

**Comment by browniefed** [comment]
@kentcdodds I get it and I suppose people familiar with glamorous will. However there is no such thing as `view` in React Native. There is only `View`. I'm afraid that people will also opt for `glamorous.View` and be confused

**Comment by kentcdodds** [comment]
I have thoughts around this but need a keyboard... 😀 Will be back later today

**Comment by kkemple** [comment]
> Also just realized. You should use Animated. Otherwise these components won't be able to be animated.

Can't users wrap a RN glamorous component in `Animated.createAnimatedComponent()`? In user land? That would mean no changes to proposed API and no bloat.

**Comment by kkemple** [comment]
> I get it and I suppose people familiar with glamorous will. However there is no such thing as view in React Native. There is only View. I'm afraid that people will always opt for glamorous.View and be confused

I agree here. Moving to lowercase would probably cause a lot of confusion.

**Comment by browniefed** [comment]
@kkemple No. The `style` hasn't been processed yet .
You have a layer of
```
-styles passed in
-- animated interpreter
--- styles processed
---- View
```
Without putting the `Animated` stuff at the lowest level closest to the view we won't be able to utilize the full power of Glamorous.

**Comment by kentcdodds** [comment]
Ok, so here's my perspective:

> I would actually be hesitant to export anything name Animated from Glamorous.

I'm not talking about adding an `Animated` export. I'm talking about monkey-patching `Animated` onto the default export. So you wouldn't be able to `import {Animated} from 'glamorous-native'` you'd have to do: `import glamorous from 'glamorous-native'; glamorous.Animated.etc...`

But in any case, it would be fine to not even worry about this I think because people could just do this:

```javascript
import glamorous from 'glamorous-native'
import Animated from 'wherever'
glamorous(Animated.View)({/* styles */})
```

I don't think that's all that bad 🤷‍♂️ 

> However there is no such thing as view in React Native. There is only View. I'm afraid that people will always opt for glamorous.View and be confused

The important distinction is that `glamorous.div` or `glamorous.view` would not be a component at all. It's a _function_ (not a constructor) that returns a component, so you could do:

```javascript
const MyStyledView = glamorous.view({/* styles */})
```

And then you could use the capital like this:

```jsx
<glamorous.View
  minHeight={32}
  padding={10}
>
  whatever
</glamorous.View>
```

If the `glamorous.View` thing is weird for you, then you can do:

```jsx
const {View} = glamorous
<View
  minHeight={32}
  padding={10}
>
  whatever
</View>
```

And if `View` conflicts with the real `View`, you could swap that easily.

```jsx
const {View: StyledView} = glamorous
<StyledView
  minHeight={32}
  padding={10}
>
  whatever
</StyledView>
```

My point is that we capitalize components and lowercase functions. I think that's a pretty widely accepted convention, and something people should understand about `glamorous` or `glamorous-native`.

**Comment by browniefed** [comment]
@kentcdodds Gotcha. I understand and generally agree with all of that. 
Except for the inability to import specific pieces and always forcing you to import glamorous as default.

A standard practice is to `import { View, XYZ ABC } from "react-native"`. I feel like this is a pattern we should attempt to maintain.

I guess the `glamorous(Animated.View)({/* styles */})` is fine. Just might get repetitive is all. Mostly this is all bike shedding on my part haha. I don't have any real issues with the API at the moment, mostly just pointing out general practices in the RN world. Also can't speak to the API since I haven't actually tested this out in any sort of application at scale.

**Comment by kentcdodds** [comment]
Your perspective is appreciated :+1:

We're actually working on some things on the web side to make it so you could import specific elements from glamorous. Add tree-shaking to that and glamorous becomes pretty small :) Learn more of the ideas in #42.

**Comment by kkemple** [comment]
> My point is that we capitalize components and lowercase functions. I think that's a pretty widely accepted convention, and something people should understand about glamorous or glamorous-native.

yeah that seems reasonable to me!

> I'm not talking about adding an Animated export. I'm talking about monkey-patching Animated onto the default export. So you wouldn't be able to import {Animated} from 'glamorous-native' you'd have to do: import glamorous from 'glamorous-native'; glamorous.Animated.etc...

I actually think this is the right way to go. It matches the RN API...
```jsx
import {Animated} from 'react-native'
import glamorous from 'glamorous-native'

export default () => (
  <Animated.View {...props} >
    <glamorous.Animated.View {...props} />
  </Animated.View>
)
```

**Comment by Andreyco** [comment]
This seems super weird to me:

```js
<glamorous.Animated.View {...props} />
```

I would rather do following:
```js
const MyView = glamorous(Animated.View)
```

**Comment by kkemple** [comment]
I believe API for web supports both. The former being the convenience method.

The first example is similar to web behavior now...
`<glamorous.Div {...props} />` so in RN that would translate to `<glamorous.View {...props} />`

Where web and RN differ is that RN has specific Animated components... so you need a way to provide those as well on top of glamorous componentFactory method.

**Comment by ssomnoremac** [comment]
> However there is no such thing as view in React Native. There is only View. I'm afraid that people will always opt for glamorous.View and be confused

I agree, especially as considering the Factory is likely the more often used API. So how do we distinguish with the JSX `<glamorous.View props={}/>` ?

**Comment by kentcdodds** [comment]
What of my point though?

> My point is that we capitalize components and lowercase functions. I think that's a pretty widely accepted convention, and something people should understand about glamorous or glamorous-native.

**Comment by kkemple** [comment]
Yeah, I am in agreement with @kentcdodds. I think `glamorous.view()` works. It matches the function vs class|component naming convention.

```jsx
<glamorous.Div />
<glamorous.View />

const Div = glamorous.div()
const View = glamorous.view()
```

I'm starting to think breaking that pattern would be way more confusing to anyone using glamorous on web...

**Comment by ssomnoremac** [comment]
`glamorous.div` corresponds to `<div>`
`glamorous.View` corresponds to `<View>`

is how I think of it. It think it would only further confuse web people. They might start writing `<view>` in JSX. But it wouldn't be the end of the world for me 😄 . I only do RN development though so I have a bias to what seems natural to a RN developer not doing any web.

**Comment by kentcdodds** [comment]
So would you just _not_ support this: `<glamorous.View />`? I just think that would be a mistake... That's one part of the API that people say gets them to even be willing to look at glamorous.

**Comment by kkemple** [comment]
I think the two environments differ enough that anyone coming from web is going to have to do some adjusting as is. I think keeping glamorous API consistent is :key:

**Comment by ssomnoremac** [comment]
There's probably a way to still support JSX-- but the convention for this library is to use letter-case to do that so I will yield to the majority here. I'll add the lowercase conversions as they exist on the web `index.js` to my fork

**Comment by kentcdodds** [comment]
Thanks for doing that. I truly believe that it's the right decision for the community. I think that `glamorous.view` as lower case makes just as much sense as any other HOC function (like `connect`) because that's exactly what it is :) I expect/hope that it wont cause as much confusion for people as has been speculated.

**Comment by kkemple** [comment]
@ssomnoremac let me know if you want any help on that, as an RN user I would love to help bring this to life :smile:

**Comment by ssomnoremac** [comment]
Sure, would love that. If you want to take a crack at adding back themes
and the rest of the API that's applicable, go for it. I'll focus on the
function/JSX distinction just discussed. Or fine if you beat me to that! My
fork is in this thread. On phone now.

**Comment by kkemple** [comment]
I will focus on Theming and HOC support. :smile:

**Comment by ssomnoremac** [comment]
Hey @kkemple , how are you progressing with the above? It's been over a week -- want to keep the momentum going on this.

**Comment by kkemple** [comment]
Yeah I was traveling and at react amsterdam so i didn't have any time, feel free to pass on to someone else or tackle yourself if you want to. I do plan to start on it but I don't want to hold you up and it seems you want a speed which i don't have the bandwidth to deliver so i'm fine with whatever. :smile:

**Comment by ssomnoremac** [comment]
I don't really want to touch theming but will push forward with other items. Thanks for letting me know.

**Comment by ssomnoremac** [comment]
Got the uppercase JSX/ lowercase factory function working [here](https://github.com/hyrwork/glamorous/tree/feature/react-native), but ran into something when testing. Changing style props on the fly with hot-reloading won't result in changes bring seen. That's because those props were cached in the component instance. Any ideas for a work-around? I'd hate to have to memoize on the stringified styles just for a development feature.

**Comment by Andreyco** [comment]
Do not memoize for dev at all? Worth trying how the performance is.

**Comment by ssomnoremac** [comment]
Yeah, I have no idea what impact memorizing the stringified styleProps (as the key) would have in terms of performance but I'd have to imagine it would take a bit of a hit. I have no idea. Wouldn't know how to test either. Could use help there.

**Comment by ssomnoremac** [comment]
could just switch on this. 
```
if (__DEV__) {
    // momoize
} else {
    // cache on instance
}
```
Never done this before. Does it work?

**Comment by ssomnoremac** [comment]
k, I've tested [this](https://github.com/hyrwork/glamorous/tree/feature/react-native) a bit (assuming `__DEV__` is falsey in production) and it seems to work. @kentcdodds have you given more thought as to if this needs to exist as it's own project?

Items still left to implement
1. Theming
2. Account for Animated components
3. Make a more comprehensive style list. Right now its just `View` and `Text` styles.
4. Refactor and cleanup a bit
5. README

**Comment by kentcdodds** [comment]
> have you given more thought as to if this needs to exist as it's own project?

It seems to be significantly different enough to merit its own project. I hope that we can keep the APIs similar if not the exact same.

Thanks for all the work on it!

**Comment by ajwhite** [comment]
Could we inventory the significant differences? From my understanding so far:
- Core components (`View` vs `div`)
- Style caching (`StyleSheet` library in ReactNative)

Is there anything in addition to this list? And how invasive is this to each module of the project? I wonder if we could decouple these platform differences from the core pieces of this project so the core logic and pieces can be as agnostic as possible.

**Comment by kentcdodds** [comment]
Thanks for chiming in @ajwhite! If anyone wants to push that effort, I'd be happy to review PRs and things. I don't have any cycles to push that effort myself I'm afraid.

**Comment by ssomnoremac** [comment]
@ajwhite those are the core differences in just `index.js` (or the now refactored version of what that file was) but we're talking about a completely different `should-forward-property.js` and `Examples`.
I think Kent is right, it needs to be mirror project. At the end of the day, I don't think the web users are going to appreciate having the extra bloat added to this project. If I take this to another domain I'd love to hear that people will support. Or if someone wants to try and keep it here, do it. But I have an immediate need for this so it would be a lot easier not to have to debate changes with web people.

**Comment by kentcdodds** [comment]
I support whatever you want to do honestly. I have no personal need for a React Native version so I really shouldn't have an opinion at all. What I would ask though is that the API is as close as possible as far as it's reasonable to do in Native. The whole learn once-write anywhere thing you know?

Thanks for your work on it!

I'll be happy to close this issue as soon as we have a PR in the README of this project to reference the React Native version :+1:

**Comment by ssomnoremac** [comment]
Thanks @kentcdodds , I'll get something set up and see how it goes, then get back to you with that PR. Last thing though, name?
glamorous-native ?

**Comment by kentcdodds** [comment]
glamorous-native makes the most sense to me, unless we can think of something more clever ;-)

**Comment by ajwhite** [comment]
Being eager, I put together a starting point so I can start using these concepts in the meantime. And when support comes for native, that it'll be a 1-1 transition.

For those that are also eager, this should get you the basics going:

```js
const createStyledComponent = (StyledComponent) => (staticStyles, dynamicStyles) => {
  const styles = StyleSheet.create({staticStyles});
  return (props) => (
    <StyledComponent {...props} style={[styles.staticStyles, dynamicStyles(props), props.style]} />
  );
};
```
```js
const StyledView = createStyledComponent(View)({
  // static styles
  height: 100,
  width: 100
}, (props) => ({
  // dynamic styles
  backgroundColor: props.active ? 'blue' : 'red'
}));
```
```js
<StyledView active={this.state.active} />
```

My main goal here is to introduce dynamic styles, this lacks any other features that glamorous provides. It's a very naive approach, no caching or memoization, just a higher order component to allow for dynamic styles.

**Comment by ssomnoremac** [comment]
@ajwhite I already have something [here](https://github.com/hyrwork/glamorous/blob/feature/react-native/src/index.js) that's pretty far along using much the same construction as web project.

**Comment by ssomnoremac** [comment]
@ajwhite you do implement the functional style rules which is missing from my fork. It's much easier to follow with the way you do it actually. The static/dynamically treated styles and the StyleSheet api is just a bit too different to try to keep to the way things are organized here. Thanks for that realization. Will get the separate project up soon.

**Comment by ssomnoremac** [comment]
Ok, I created the [project](https://github.com/hyrwork/glamorous-native). Obviously bare bones, but we need to figure out some things first. Really, the hardest part is that styles can come flying in from every direction in a ton of different ways and the component needs to be able to separate dynamic ones from static. I created a diagram to help clear that up. Can we start the discussion there?

**Comment by kentcdodds** [comment]
Terrific! 🎉 I've created https://github.com/hyrwork/glamorous-native/issues/1 to continue discussion from here.

Thanks for all the input everyone!

**Comment by ajwhite** [comment]
Shoot! Apologies I hadn't shared this earlier -- we've been heads down and figured this was still in an early experimental stage that wasn't yet ready for establishing a home. 🙇 

Our team had been experimenting in parallel due to an immediate need/want to build a new product feature set with this library in React Native in our latest sprint. We started small with a basic implementation in our project a few weeks ago (shared recently in https://github.com/paypal/glamorous/issues/7#issuecomment-300486442) to unblock building new features with the `glamorous` pattern, and in the recent weeks we began fleshing out the rest of the features to use. 

I actually feel rather guilty writing this with the issue closed, as it adds overlap to others work and you guys have also been thinking about this port; communication on my part could have been better -- apologies this is coming in last minute! It can be a challenge to share things while they are part of a private codebase governed by certain rules on access and sharing. 

I'm almost done extracting this from the codebase and into a public repo (ETA this evening). This port is feature and parity complete with `glamor`, with units and has been battle tested in some of our release channels. Basically good to go aside from documentation and CI - as that was integrated in the previous project. Would you consider bringing all options back onto the table? I'm sure we could also incorporate efforts that have already been made.

**Comment by kentcdodds** [comment]
I think that we can all agree that if the implementation is better and still honors the `glamorous` API (or only diverts for great reasons) that we'd love to see what you have. That is to say, all options are always on the table because this is open source and that's the way things work :) I don't do React Native, but I'm sure many look forward to seeing what you've got!

**Comment by ajwhite** [comment]
Hey folks!

Excited to share [robinpowered/glamorous-native](https://github.com/robinpowered/glamorous-native) 🌟 

This is a port that we've been using internally, and given the attention and requests for `react-native`, we're bringing it to you. 

You'll find the project is completely 1-1 with `glamorous`, the only differences being some particular private apis (that differ between `glamor` and React Native's `StyleSheet`), and the building block components (`<glamorous.Div />` vs `<glamorous.View />`). Otherwise, the rest of the API is honored 👍  This also ships with the [tiny](https://github.com/robinpowered/glamorous-native/blob/master/src/tiny.js) module.

I tried to keep the private APIs and the project itself structured similarly to `glamorous`. The intention being that there's already a strong community behind this, and I'd like it to be as simple as possible to transition efforts over to `react-native` for interested folks.



---

[Implementation files](https://github.com/robinpowered/glamorous-native/tree/master/src) can be navigated both at the directory/file level, as well as within the code, similarly to how you would navigate through `glamorous`.

[Unit tests](https://github.com/robinpowered/glamorous-native/tree/master/src/__tests__) are 1-1 as well, with the only difference being the base component being used to test. You'll find the `glamorous` test suit reimplemented and passing in `glamorous-native`. There are about 5 tests I still need to work on (**contributions welcome**), as I'm having a tricky time using `mount` with `react-native` to verify certain workflows that are tested through multiple stages of a component lifecycle. Code Coverage is at 91%.

[An example](https://github.com/robinpowered/glamorous-native/tree/master/examples) can be pulled down and ran directly if you'd like to give it a go 

[The issues](https://github.com/robinpowered/glamorous-native/issues) are populated with `good first task` stories -- particularly around performance and preparing for its first potential release.

[The build environment](https://github.com/robinpowered/glamorous-native/blob/master/package-scripts.js) is as close to `glamorous` as possible, making use of `nps` and aiming to follow the https://github.com/robinpowered/glamorous-native/blob/master/.travis.yml) currently runs the validation suite and reports coverage.

[Community resources](https://github.com/robinpowered/glamorous-native/tree/master/.github) have been set up similarly as they are in this project, along with [contribution guidelines](https://github.com/robinpowered/glamorous-native/blob/master/CONTRIBUTING.md)

### What can you do?

It would be great if interested folks could poke around, and if you've implemented something you don't see in this project, feel free to open up a pull request! You're likely familiar with the `glamorous` codebase, so this will be familiar territory.

Otherwise tomorrow I'll be tying up loose ends, mainly in the build and release flow 🚀

**Comment by kentcdodds** [comment]
Wow! This is fantastic! Thanks for putting all that work in! Let's see how we can convince efforts to make this as good as it can be!

**Comment by kentcdodds** [comment]
Oh, and let me know when we've decided which direction to go with an official implementation and we can link it in the README and I'll tweet it out 💪

**Comment by ajwhite** [comment]
@kentcdodds I'm just about ready to tag a `v1.0.0`, have some folks touching up documentation and the examples. We've set up a [list of issues](https://github.com/robinpowered/glamorous-native/issues) that would be great first tasks for folks to continue optimizing the project. There's opportunities for folks that have been doing existing work, particularly in optimizations, that can combine efforts.

I'd like to host this project, we have a dedicated team behind it and we share the same values discussed in this thread -- keeping the APIs exactly how they are in this project. We see `glamorous` as a parent project to a complimentary `react-native` version. Any changes to behavior/interfaces here will be maintained `glamorous-native`. If proposals are introduced on `glamorous-native`, they'd be shared with `glamorous` and introduced if it appears to be a joint agreement (if it's a react-native specific thing, that's a little different ofcourse).

**Comment by kentcdodds** [comment]
I would love to hear from other folks on this thread. As a non-react-native user, I'm unsure whether I can say that this solution is the one we should run with. Just would like others to review it. Especially @ssomnoremac who's put a bit of effort into this.

I have no problem with you hosting the project yourself. I definitely want to give you that autonomy. I'll probably not have a lot to do with the project myself. Ya'll seem to be doing great work already!

**Comment by ajwhite** [comment]
Sounds good - for others in this thread, the [README](https://github.com/robinpowered/glamorous-native#readme) is a good place to begin an overview.

To preface - if we consider the lower level React Components and DOM elements as "primitive components", `glamorous` and `glamorous-native` both act as HOC to these.

`glamorous` and `glamorous-native` differ in the following areas:
- The [built-in component factories](https://github.com/robinpowered/glamorous-native/blob/707b6c5c1c63a819bce7a9512af0123a7fc1d506/src/index.js#L11-L18) - `glamorous.div()` vs `glamorousNative.view()`
- React Native component [property and style property forwarding detection](https://github.com/robinpowered/glamorous-native/blob/master/src/is-style-prop.js). This uses React Native's component style prop types to determine if styles are supported. There's still room for improvement outlined in https://github.com/robinpowered/glamorous-native/issues/9 and https://github.com/robinpowered/glamorous-native/issues/10

Aside from that, you folks should see parity between the two. Happy to answer any questions, and please open an issue or PR!

**Comment by kentcdodds** [comment]
Again as a non-react native user, my favorite part of glamorous-native is the logo:

<img width="449" alt="screen shot 2017-05-16 at 10 47 41 pm" src="https://cloud.githubusercontent.com/assets/1500684/26168518/8dba5740-3af7-11e7-8512-30acc432991e.png">

It's just so awesome.

**Comment by ajwhite** [comment]
Also -- highly encourage folks to give the following examples a run through to play around

- [basic ReactNative project](https://github.com/robinpowered/glamorous-native/tree/master/examples/react-native-glamorous)
- [ReactNative project w/ `glamorous` themes](https://github.com/robinpowered/glamorous-native/tree/master/examples/react-native-glamorous-themes)

**Comment by tabrindle** [comment]
Been watching this thread for a while - we've been using styled components with react native and like the concept but had the same issues that lead @kentcdodds to writing glamourous. @ajwhite we're getting started with glamourous-native now, and will let you know how it goes!

**Comment by ajwhite** [comment]
Sounds great, @tabrindle - I've updated the [README](https://github.com/robinpowered/glamorous-native#readme) to contain [getting started instructions](https://github.com/robinpowered/glamorous-native/commit/ee68b2ce548bd5396768c4fbebfdab9d8247c62d) for people trying out the pre-release.

**Comment by ssomnoremac** [comment]
Hey @ajwhite , was away much of the day and wow, yeah, looks great! I can start using this right away and let you know how it goes. In the end it'll perhaps be better maintained with the support of the Robin team than with our current, earlier-stage company. Will chime in about caching, animation most likely. Excited to continue the discussions in issues over there!

**Comment by ajwhite** [comment]
Great @ssomnoremac, a few areas that will be of interest to you:

### Caching

The [first thing that happens](https://github.com/robinpowered/glamorous-native/blob/b9629634e640c9bd4cb973045c551d74de7e8c81/src/create-glamorous.js#L29) when a `glamorousComponentFactory` method is called is to [create `StyleSheet` registrations for in-line style objets](https://github.com/robinpowered/glamorous-native/blob/b9629634e640c9bd4cb973045c551d74de7e8c81/src/create-glamorous.js#L15).

For example:
```js
const StyledView = glamorous.view({backgroundColor: 'red'})
```
Will result in creating a registered style (`StyleSheet.create`), as these styles can never change.

When `StyledView` becomes rendered, it [takes the styles provided from the factory call, along with any style overrides and the theme, and [builds a final style rule](https://github.com/robinpowered/glamorous-native/blob/b9629634e640c9bd4cb973045c551d74de7e8c81/src/create-glamorous.js#L87-L92) for the rendered component.

That rule is created in [src/get-styles.js](https://github.com/robinpowered/glamorous-native/blob/master/src/get-styles.js)

This simply creates an array of the styles, similar to `styles={[foo, bar, hello, world]}`.

I've been thinking about doing something with `StyleSheet.flatten` and caching the resulting value (which would merge that array into a single object). Downside - registered `StyleSheet` styles are now back to plain objects. More on this in https://github.com/robinpowered/glamorous-native/issues/5.

### Animated

Haven't had the chance to look at or experiment with this, great to hear you're interested in that. https://github.com/robinpowered/glamorous-native/issues/17 is where the story currently lives.

**Comment by ajwhite** [comment]
Hey @tabrindle and @ssomnoremac - looking forward to hearing how things went 😃 

I'd love to ship this during ReactEurope given the attention to new stories, features, and developments surrounding React. Would love a +1 from you guys if you're satisfied with the project. 

A couple updates since yesterday:
- I've confirmed `Animated` works terrifically! See my update in https://github.com/robinpowered/glamorous-native/issues/17#issuecomment-302418531
- A blog post is drafted for the announcement, feel free to read over, give feedback, and tell me if it's too similar to glamorous' blog post. I kept the format similar since, well, the projects are the same! https://medium.com/@atticoos/introducing-glamorous-for-react-native-1b8365e7f33f
- Continue'd to polish documentation and prepare for the initial release

On my end, `v1.0.0` is ready, and I'd love to push this today! 🙌 

Thank you everybody.
